### PR TITLE
[LV][VPlan] Introduce VPScalarStore recipe to handle scalar stores in exit block.

### PR DESCRIPTION
**Pull request by Mel-Chen**
A new recipe, VPScalarStore, to describe and handle scalar stores that sink into the exit block. This patch separates the computation of the final reduction result and the generation of intermediate stores.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 1]
```diff
@@ -2821,6 +2823,55 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags {
   }
 };
 
+/// A recipe to represent scalar stores that sink outside the vector loop.
+class VPIntermediateStoreRecipe : public VPRecipeBase {
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlan.h:2827]: Could this simply be a VPInstruction opcode?

File: llvm/test/Transforms/LoopVectorize/ARM/tail-fold-multiple-icmps.ll [hunk 2]
```diff
@@ -16,41 +16,41 @@ define arm_aapcs_vfpcc i32 @minmaxval4(ptr nocapture readonly %x, ptr nocapture
 ; CHECK-NEXT:    br label [[VECTOR_BODY:%.*]]
 ; CHECK:       vector.body:
 ; CHECK-NEXT:    [[INDEX:%.*]] = phi i32 [ 0, [[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], [[VECTOR_BODY]] ]
-; CHECK-NEXT:    [[VEC_PHI:%.*]] = phi <4 x i32> [ <i32 2147483647, i32 2147483647, i32 2147483647, i32 2147483647>, [[VECTOR_PH]] ], [ [[TMP3:%.*]], [[VECTOR_BODY]] ]
-; CHECK-NEXT:    [[VEC_PHI1:%.*]] = phi <4 x i32> [ <i32 -2147483648, i32 -2147483648, i32 -2147483648, i32 -2147483648>, [[VECTOR_PH]] ], [ [[TMP2:%.*]], [[VECTOR_BODY]] ]
+; CHECK-NEXT:    [[VEC_PHI:%.*]] = phi <4 x i32> [ <i32 2147483647, i32 2147483647, i32 2147483647, i32 2147483647>, [[VECTOR_PH]] ], [ [[TMP2:%.*]], [[VECTOR_BODY]] ]
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/ARM/tail-fold-multiple-icmps.ll:19]: unrelated test changes?

File: llvm/test/Transforms/LoopVectorize/interleaved-accesses.ll [hunk 3]
```diff
@@ -108,23 +108,23 @@ define void @test_struct_array_load3_store3() {
 ; CHECK-NEXT:    br label [[VECTOR_BODY:%.*]]
 ; CHECK:       vector.body:
 ; CHECK-NEXT:    [[INDEX:%.*]] = phi i64 [ 0, [[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], [[VECTOR_BODY]] ]
-; CHECK-NEXT:    [[TMP0:%.*]] = mul i64 [[INDEX]], 12
-; CHECK-NEXT:    [[NEXT_GEP:%.*]] = getelementptr i8, ptr @A, i64 [[TMP0]]
+; CHECK-NEXT:    [[OFFSET_IDX:%.*]] = mul i64 [[INDEX]], 12
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/interleaved-accesses.ll:111]: unrelated test changes?

File: llvm/test/Transforms/LoopVectorize/interleaved-accesses.ll [hunk 4]
```diff
@@ -266,21 +266,21 @@ define void @test_struct_store4(ptr noalias nocapture readonly %A, ptr noalias n
 ; CHECK-NEXT:    br label [[VECTOR_BODY:%.*]]
 ; CHECK:       vector.body:
 ; CHECK-NEXT:    [[INDEX:%.*]] = phi i64 [ 0, [[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], [[VECTOR_BODY]] ]
-; CHECK-NEXT:    [[TMP0:%.*]] = shl i64 [[INDEX]], 2
-; CHECK-NEXT:    [[NEXT_GEP:%.*]] = getelementptr i8, ptr [[A:%.*]], i64 [[TMP0]]
+; CHECK-NEXT:    [[OFFSET_IDX:%.*]] = shl i64 [[INDEX]], 2
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/interleaved-accesses.ll:269]: unrelated test changes?

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 1 above)
Comment (nikolaypanchenko) [reply llvm/lib/Transforms/Vectorize/VPlan.h:2827]: I believe the way Mel implemented it is similar to how `VPWidenStore` is done, with the difference that it's scalar store. Maybe proper name would be `VPScalarStore` though, like `VPScalarCast`

File: llvm/test/Transforms/LoopVectorize/ARM/tail-fold-multiple-icmps.ll (hunk 2 above)
Comment (Mel-Chen) [reply llvm/test/Transforms/LoopVectorize/ARM/tail-fold-multiple-icmps.ll]: c7ec02de4da1a16f33efbbf25ebb84cb676731f9
My bad, removed them.

File: llvm/test/Transforms/LoopVectorize/reduction-with-invariant-store.ll [hunk 5]
```diff
@@ -625,10 +625,10 @@ exit:
 define void @reduc_mul_add_store_same_ptr(ptr %dst, ptr readonly %src) {
 ; CHECK-LABEL: define void @reduc_mul_add_store_same_ptr
 ; CHECK:       middle.block:
-; CHECK-NEXT:    [[TMP4:%.*]] = call i32 @llvm.vector.reduce.add.v4i32(<4 x i32> [[TMP3:%.*]])
-; CHECK-NEXT:    store i32 [[TMP4]], ptr %dst, align 4
-; CHECK-NEXT:    [[TMP2:%.*]] = call i32 @llvm.vector.reduce.mul.v4i32(<4 x i32> [[TMP1:%.*]])
-; CHECK-NEXT:    store i32 [[TMP2]], ptr %dst, align 4
```
Comment (nikolaypanchenko) [inline llvm/test/Transforms/LoopVectorize/reduction-with-invariant-store.ll:631]: I wonder why this was the correct order of stores

Comment (Mel-Chen) [reply llvm/test/Transforms/LoopVectorize/reduction-with-invariant-store.ll:631]: ...That's a good question. I think this was the incorrect order of stores. 
@fhahn @igogo-x86 Could you help verify it again?

**Comment by Mel-Chen** [comment]
Ping

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 6]
```diff
@@ -1927,6 +1921,24 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
 }
 #endif
 
+void VPScalarStoreRecipe::execute(VPTransformState &State) {
+  IRBuilderBase &Builder = State.Builder;
+  Value *StoredVal = State.get(getStoredVal(), 0, /*IsScalar*/ true);
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1926]: if it is sunk outside, shouldn't it store the last lane of the last unrolled iteration? as is, it would generate incorrect code when interleaving with other users?

Or alternatively enforce that the operand are also defined outside any vector region?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 7]
```diff
@@ -2963,6 +2965,54 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags {
   }
 };
 
+/// A recipe to represent scalar stores that sink outside the vector loop.
+class VPScalarStoreRecipe : public VPRecipeBase {
+  StoreInst &SI;
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlan.h:2970]: Can this be removed and be replaced by fields that store the needed information independent of StoreInst, so it can also be created without underlying IR?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 8]
```diff
@@ -2963,6 +2965,54 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags {
   }
 };
 
+/// A recipe to represent scalar stores that sink outside the vector loop.
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlan.h:2968]: Would be good to remove the reference of sinking, the main thing is it should be outside any vector loop region an will only generate a single instance

File: llvm/test/Transforms/LoopVectorize/reduction-with-invariant-store.ll (hunk 5 above)
Comment (fhahn) [reply llvm/test/Transforms/LoopVectorize/reduction-with-invariant-store.ll:631]: New order looks correct, storing the mul before the add, as in the original IR

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: In general, VPInstruction is intended to be used for cases where Opcode + VPVAlue operands are sufficient to define the operation to avoid explosion of recipe classes. There are other opcodes that don't define values (this is similar to LLVM IR's Instructions which also all are Values, but possibly of VoidTy.

In this case, it looks like it would be good to store additional data (alignment), so having a separate recipe seems fine.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 6 above)
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: ee8ce9f65a27313977403fd74f84292bac40d7ea
Check operands in the ::execute function. 
Initially, I attempted to check the operand in the constructor, but given the current application, the stored value is only correctly assigned in FixReduction.

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 7 above)
Comment (Mel-Chen) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: 818c8e3bb7420571d4dd2f3fbdfc2ecc1e1eee31
Done, but I can't imagine how it could be used so far. That would be great If you're willing to share your plan.

**Comment by Mel-Chen** [comment]
Ping :)

**Comment by Mel-Chen** [comment]
Ping

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 9]
```diff
@@ -8911,8 +8882,9 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
       for (VPUser *U : Cur->users()) {
         auto *UserRecipe = dyn_cast<VPSingleDefRecipe>(U);
         if (!UserRecipe) {
-          assert(isa<VPLiveOut>(U) &&
-                 "U must either be a VPSingleDef or VPLiveOut");
+          assert((isa<VPLiveOut>(U) || isa<VPScalarStoreRecipe>(U)) &&
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:8885]: better to also assert that `VPScalarStoreRecipe` is outside the vector loop region?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 10]
```diff
@@ -9130,8 +9102,9 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
         VPInstruction::ComputeReductionResult, {PhiR, NewExitingVPV}, ExitDL);
     FinalReductionResult->insertBefore(*MiddleVPBB, IP);
     OrigExitingVPV->replaceUsesWithIf(
-        FinalReductionResult,
-        [](VPUser &User, unsigned) { return isa<VPLiveOut>(&User); });
+        FinalReductionResult, [](VPUser &User, unsigned) {
+          return isa<VPLiveOut>(&User) || isa<VPScalarStoreRecipe>(&User);
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9106]: better to also assert that VPScalarStoreRecipe is outside the vector loop region?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 11]
```diff
@@ -1102,6 +1103,7 @@ class VPRecipeWithIRFlags : public VPSingleDefRecipe {
 public:
   static inline bool classof(const VPRecipeBase *R) {
     return R->getVPDefID() == VPRecipeBase::VPInstructionSC ||
+           R->getVPDefID() == VPRecipeBase::VPScalarStoreSC ||
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlan.h:1106]: VPScalarStoreRecipe doesn't inherit from VPRecipeWithIRFlags?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 12]
```diff
@@ -2964,6 +2966,60 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags {
   }
 };
 
+/// A recipe to represent scalar stores that are outside the vector loop region.
+/// Both the stored value and the address must be uniform, and finally, only a
+/// single store instance will be generated.
+class VPScalarStoreRecipe : public VPRecipeBase {
+  /// The alignment information of this store.
+  MaybeAlign Alignment = std::nullopt;
+  /// The ingredient store of this recipe.
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlan.h:2975]: would be good to clarify that this is only used to access metadata?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 13]
```diff
@@ -1927,6 +1921,29 @@ void VPScalarCastRecipe ::print(raw_ostream &O, const Twine &Indent,
 }
 #endif
 
+void VPScalarStoreRecipe::execute(VPTransformState &State) {
+  [[maybe_unused]] bool IsAllOperandsUniformAfterVectorization =
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1925]: assert that both the operand and the recipe itself are defined outside the vector loop region, otherwise we might get unexpected results?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 14]
```diff
@@ -2964,6 +2966,60 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags {
...
+
+  /// Return the value stored by this recipe.
+  VPValue *getStoredVal() const { return getOperand(0); }
+
+  /// Return the address accessed by this recipe.
+  VPValue *getAddress() const { return getOperand(1); }
+
+  /// Returns true if the recipe only uses the first lane of operand \p Op.
+  bool onlyFirstLaneUsed(const VPValue *Op) const override {
+    assert(is_contained(operands(), Op) &&
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlan.h:3017]: is this needed? operands at the moment are only operands are ComputeReductionReusult which produces a single scalar anyways?

**Comment by Mel-Chen** [comment]
Closed since we have a way without adding new recipe.
Redirect to [#106342](https://github.com/llvm/llvm-project/pull/106342)
